### PR TITLE
Enhance parse_user_data.py resiliency

### PR DESCRIPTION
**Pull request by debmc**
Check for value of None returned from parsers (json.loads crashes) and return a helpful
message to help further identify any issue.

RTC: 320461
Signed-off-by: Deb McLemore <debmc@us.ibm.com>

### CODE REVIEW

File: modules/pel/peltool/user_data.py [hunk 1]
```diff
@@ -37,6 +37,10 @@ def toJSON(self) -> OrderedDict:
 
         value = parser.parse()
 
+        # Catch if value is None, otherwise python crashes
+        if value is None:
```
Comment (spinler) [inline modules/pel/peltool/user_data.py:41]: A few things:

The ExtendedUserData section also parses user data, that's why the ParseUserData class up above is used, so the code that interacts with the plugins is just in one place.  So any changes would be done in that class.

Also the changes here wouldn't even print a hex dump of the data, which I think would be better than nothing.

I think you could do something similar to this error path that is already there:

            print('Failed parsing user data for creator {} compID {} '
                  'subType {} version {}: {}'.format(
                      self.creatorID, "0x%04X" % self.compID,
                      "0x%X" % self.subType, self.version, str(e)), file=sys.stderr)
            mv = memoryview(self.data)
            return json.dumps(hexdump(mv))

if you wanted, I suppose you could add the error into the data along with the hexdump.  'plugin returned None' or something like that being the message.

Comment (debmc) [reply modules/pel/peltool/user_data.py:41]: Thanks, I'll setup a recreate of the error paths and see how this can be modified.

File: modules/pel/peltool/parse_user_data.py [hunk 2]
```diff
@@ -40,25 +40,35 @@ def parse(self) -> str:
         else:
             value = self.parseCustom()
 
+        # Catch if value is None, otherwise python crashes
+        if value == None:
+            value = ("PROBLEM with UNRECOGNIZED value from parser creatorID={} compID={} subType={} version={}"
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:45]: could just say 'Parser returned a value of None' so it says what the real problem is.

File: modules/pel/peltool/parse_user_data.py [hunk 3]
```diff
@@ -40,25 +40,35 @@ def parse(self) -> str:
...
+                     .format(self.creatorID, hex(self.compID), self.subType, self.version))
+            return json.dumps(value)
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:47]: you still aren't returning the hexdump of the original data...

Could something like the following to get both:
output['Error'] = "Plugin returned a value of None..."
mv = memoryview(self.data)
output['Original Data'] = hexdump(mv)
return json.dumps(output)

File: modules/pel/peltool/parse_user_data.py [hunk 4]
```diff
@@ -40,25 +40,35 @@ def parse(self) -> str:
...
         return value
 
     def parseCustom(self) -> str:
         import importlib
         name = (self.creatorID.lower() + "%04X" % self.compID).lower()
         try:
             cls = importlib.import_module("udparsers." + name + "." + name)
-            mv = memoryview(self.data)
-            return cls.parseUDToJson(self.subType, self.version, mv)
+            if self.data:
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:55]: this would be protecting against an empty UserData section.  Is that what you saw?

File: modules/pel/peltool/parse_user_data.py [hunk 5]
```diff
@@ -40,25 +40,35 @@ def parse(self) -> str:
...
+                mv = memoryview(self.data)
+                return cls.parseUDToJson(self.subType, self.version, mv)
         except ImportError:
-            mv = memoryview(self.data)
-            return json.dumps(hexdump(mv))
+            print("Failed parseCustom ImportError name={} creatorID={} compID={}"
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:59]: this just means there isn't a plugin registered for that UserData section, which we'd see for ever PHYP error, so I'd rather not print anything here.

File: modules/pel/peltool/parse_user_data.py [hunk 6]
```diff
@@ -40,25 +40,35 @@ def parse(self) -> str:
...
+                mv = memoryview(self.data)
+                return json.dumps(hexdump(mv))
         except Exception as e:
             print('Failed parsing user data for creator {} compID {} '
                   'subType {} version {}: {}'.format(
                       self.creatorID, "0x%04X" % self.compID,
-                      "0x%X" % self.subType, self.version, str(e)), file=sys.stderr)
-            mv = memoryview(self.data)
-            return json.dumps(hexdump(mv))
+                      "0x%X" % self.subType, self.version, str(e)))
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:68]: why did you remove printing to stderr?

File: modules/pel/peltool/parse_user_data.py (hunk 5 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:59]: Sounds good ...

File: modules/pel/peltool/parse_user_data.py (hunk 4 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:55]: Yes, on a few error flows.

File: modules/pel/peltool/parse_user_data.py (hunk 3 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:47]: The problem I was solving was that there was NO data, so the message is the ONLY available output.

File: modules/pel/peltool/parse_user_data.py (hunk 2 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:45]: Sounds good ...

File: modules/pel/peltool/parse_user_data.py (hunk 6 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:68]: By printing to stderr, the output is orphaned if the PEL output is reviewed (like reviewing PELs directed to files, etc).

Comment (spinler) [reply modules/pel/peltool/parse_user_data.py:68]: It's supposed to be.  The output is in JSON and is supposed to be machine readable.  Writing plain text will break that.

Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:68]: The print output was being routed to stderr, the change allows the stdout information be dumped to stdout, which will be captured by any re-direction.

Comment (spinler) [reply modules/pel/peltool/parse_user_data.py:68]: We can't break the JSON output to stdout, or we will crash people's tools.

How about you do what I suggested in the other spot and add the error into the JSON output along with the hex dump?

Comment (debmc) [reply modules/pel/peltool/parse_user_data.py]: Done

File: modules/pel/peltool/parse_user_data.py (hunk 3 above)
Comment (spinler) [reply modules/pel/peltool/parse_user_data.py:47]: there could be other cases where there is valid data passed in but the parser still returns None due to other reasons.

With the check you added on line 55 I don't think we'd get here with empty data again anyway.

File: modules/pel/peltool/parse_user_data.py [hunk 7]
```diff
@@ -40,25 +40,36 @@ def parse(self) -> str:
...
-                      self.creatorID, "0x%04X" % self.compID,
-                      "0x%X" % self.subType, self.version, str(e)), file=sys.stderr)
-            mv = memoryview(self.data)
-            return json.dumps(hexdump(mv))
+            d = dict()
+            # in case we do NOT have data, dump the Error at a minimum
+            if self.data:
+                mv = memoryview(self.data)
+                d["Data"] = hexdump(mv)
+            d["Error"] = ("Failed parsing user data for creator={} compID={} subType={} version={} Exception={}"
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:69]: At first this looked like an indentation error to me though it isn't.  could you move it up after line 64 to make it more obvious it should always run?

Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:69]: Will do.

File: modules/pel/peltool/parse_user_data.py (hunk 3 above)
Comment (spinler) [reply modules/pel/peltool/parse_user_data.py:47]: I think we still need to do the hexdump here because if a plugin returns None when the data isn't empty it wouldn't print any dadta.

Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:47]: Help clarify something for me ... self.data is being processed outside this logic in parse or parseCustom, for the parse leg in the normal case the self.data comes back within the value parm, correct ?

Or is the statement that similarly to the new logic added in parseCustom, we need to do the same handling as was added under the Exception flow ?

Comment (debmc) [reply modules/pel/peltool/parse_user_data.py]: Done

**Review by spinler — APPROVED** [review]

**Review by spinler — COMMENTED** [review]
the pyright linter is showing some errors for these changes.  I'd like to take a look at them first before merging.

**Review by spinler — COMMENTED** [review]
since you removed writing to stderr, the file doesn't need the 'import sys' anymore on line 7... could you remove that line?  While you're at it, could you remove lines 1 & 2 since I don't think they are needed anymore either?

File: modules/pel/peltool/parse_user_data.py [hunk 8]
```diff
@@ -40,25 +40,42 @@ def parse(self) -> str:
...
-            print('Failed parsing user data for creator {} compID {} '
-                  'subType {} version {}: {}'.format(
-                      self.creatorID, "0x%04X" % self.compID,
-                      "0x%X" % self.subType, self.version, str(e)), file=sys.stderr)
-            mv = memoryview(self.data)
-            return json.dumps(hexdump(mv))
+            d = dict()
+            # in case we do NOT have data, dump the Error at a minimum
+            d["Error"] = ("Failed parsing user data for creator={} compID={} subType={} version={} Exception={}"
+                              .format(self.creatorID, hex(self.compID), hex(self.subType), self.version, e))
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:73]: self.compID is a string, so it can't have hex() called on it.

File: modules/pel/peltool/parse_user_data.py [hunk 9]
```diff
@@ -40,25 +40,42 @@ def parse(self) -> str:
...
+            if self.data:
+                mv = memoryview(self.data)
+                d["Data"] = hexdump(mv)
+            return json.dumps(d)
+
```
Comment (spinler) [inline modules/pel/peltool/parse_user_data.py:78]: now that you added the 'if' statements on lines 61 and 66, it's possible to make it to the end of the function where there is no return statement.  I guess it could returns a 'json.dumps("")'?  or would there be something better to return there...

**Review by spinler — CHANGES_REQUESTED** [review]

File: modules/pel/peltool/parse_user_data.py (hunk 8 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:73]: Re-confirmed, self.compID is an int, so the hex formatting works.

File: modules/pel/peltool/parse_user_data.py [hunk 10]
```diff
@@ -40,25 +40,42 @@ def parse(self) -> str:
         else:
             value = self.parseCustom()
 
+        # Catch if value is None, otherwise python crashes
+        if value == None:
+            d = dict()
+            # in case we have problems, try to make every attempt to get some data points out
+            d["Error"] = ("Parser returned a value of None for creatorID={} compID={} subType={} version={}"
+                     .format(self.creatorID, hex(self.compID), self.subType, self.version))
```
[the remark this answers is not in the document]
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:48]: Re-confirmed, self.compID is an int, so the hex formatting works.

File: modules/pel/peltool/parse_user_data.py (hunk 9 above)
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py:78]: Added the json.dumps("")

File: modules/pel/peltool/parse_user_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (spinler) [reply modules/pel/peltool/parse_user_data.py:45]: Ah, what's throwing off the linter is line 25, which tries to say that compID is a string.  Think you could fix that up quick?

That is my bad.  It makes me sad that python even allows that in the first place.

[the remark this answers is not in the document]
Comment (debmc) [reply modules/pel/peltool/parse_user_data.py]: Done

**Review by spinler — APPROVED** [review]
